pkg/controller/endpoint/endpoints_controller_cache.go

```diff
@@ -0,0 +1,232 @@
+// potential race-conditions in the endpoints-controller. In addition it allows to easily determine
+// which object (pod/service) changes resulted in the endpoints object change, which is crucial in
+// implementing the network programming latency metric.
+//
+// The cache should be initialized via pod and service listers and kept up-to-date via the
+// (Add|Update|Delete)(Pod|Service) methods called in the EndpointsController's event handlers.
+// The Initialize method together with the event handlers method are guarded by the same mutex which
+// guarantees that no information is ever missed and all events are processed in the serial order.
+//
+// The cache doesn't store full objects, it stores only references to minimize the memory
+// consumption.
+type endpointsControllerCache struct {
+	pods     map[ /* namespace */ string]map[ /* name */ string]*v1.Pod
+	services map[ /* namespace */ string]map[ /* name */ string]*v1.Service
+
+	// Mutex guarding this util.
+	mutex sync.Mutex
+}
+
+// Creates new instance of the endpointsControllerCache.
+func newEndpointsControllerCache() *endpointsControllerCache {
+	return &endpointsControllerCache{
+		pods:     make(map[ /* namespace */ string]map[ /* name */ string]*v1.Pod),
+		services: make(map[ /* namespace */ string]map[ /* name */ string]*v1.Service),
+	}
+}
+
+// Initialize initializes the cache with pods and services obtained from the provided listers.
+// Pods and services are listed behind mutex, which temporarily disables event listener callbacks
+// (e.g. AddPod, UpdateService, etc.). It guarantees that no information will be lost.
+// This method should be called exactly once on the startup, before the first call of the
+// GetServiceAndPods method, after the lister have synced. Please note that this initialization
+// scheme relies on the assumption that, when a handler is registered in the Informer, the current
+// state will be fully delivered to the event handler via the OnAdd callbacks.
+// TODO(mm4tt): Support initialization natively in the Informer framework and get rid of this
+// 							method. This will require a new signal (and callback in the event handler) informing
+//              that the state has been synced and fully delivered to the event handlers via the
+//              OnAdd callbacks.
+func (e *endpointsControllerCache) Initialize(serviceLister corelisters.ServiceLister, podLister corelisters.PodLister) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	services, err := serviceLister.List(labels.Everything())
+	if err != nil {
+		// Error shouldn't happened as we're getting objects from a local cache, so fail if it does.
+		klog.Fatalf("unable to list services due to %v", err)
+	}
+	// Please note that we're listing pods at a different time than services, so it's possible that we
+	// will initialize the cache with a pod state newer than the service state. This inaccuracy
```

Comment:
That's not true. There is completely no synchronization between different object types (in this example pods and services) so there can be arbitrary difference between those (both pods lagging behind service as well as services lagging behind pods).